cassovary-core/src/main/scala/com/twitter/cassovary/graph/ArrayBasedDynamicDirectedGraph.scala

```diff
@@ -0,0 +1,206 @@
+package com.twitter.cassovary.graph
+
+import com.twitter.cassovary.graph.node.DynamicNode
+import it.unimi.dsi.fastutil.ints.IntArrayList
+import scala.collection.mutable.ArrayBuffer
+import StoredGraphDir._
+import com.twitter.cassovary.util.FastUtilUtils
+
+/**
+ * A dynamic directed graph, implemented using an ArrayBuffer of IntArrayList
+ * (from the fastutil library).  If n nodes are used, O(n) objects are created,
+ * independent of the number of edges.
+ *
+ * For efficiency, it's recommended that ids be sequentially numbered, as
+ * an internal array is stored which is longer than the maximum id seen.
+ *
+ * This class is not threadsafe.  If multiple thread access it simultaneously
```

Comment:
at least in ArrayBasedDirectedGraph, single thread performance was abysmal. From our past experience, reading graph is a time consuming operation and so it might pay to do it in multiple threads